x-pack/plugins/fleet/server/services/files/client.ts

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Readable } from 'stream';
+
+import assert from 'assert';
+
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import type { Logger } from '@kbn/core/server';
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+
+import type { FileClient } from '@kbn/files-plugin/server';
+import { createEsFileClient, createFileHashTransform } from '@kbn/files-plugin/server';
+import type { File } from '@kbn/files-plugin/common';
+
+import { v4 as uuidV4 } from 'uuid';
+
+import moment from 'moment';
+
+import { getFileDataIndexName, getFileMetadataIndexName } from '../../../common';
+
+import { FleetFileNotFound, FleetFilesClientError } from '../../errors';
+
+import type {
+  FileCustomMeta,
+  FleetFile,
+  FleetFileClientInterface,
+  FleetFileType,
+  FleetFileUpdatableFields,
+  HapiReadableStream,
+  HostUploadedFileMetadata,
+} from './types';
+
+/**
+ * Public interface for interacting with Files stored in Fleet indexes. Service is consumed via
+ * the Fleet `Plugin#start()` interface - {@link FleetStartContract}
+ */
+export class FleetFilesClient implements FleetFileClientInterface {
+  private readonly esFileClient: FileClient;
+  private readonly fileMetaIndex: string;
+  private readonly fileDataIndex: string;
+
+  constructor(
+    private esClient: ElasticsearchClient,
+    private logger: Logger,
+    packageName: string,
+    private type: FleetFileType,
+    maxSizeBytes?: number
+  ) {
+    if (!packageName) {
+      throw new FleetFilesClientError('packageName is required');
+    }
+
+    if (type === 'from-host') {
+      this.fileMetaIndex = getFileMetadataIndexName(packageName);
+      this.fileDataIndex = getFileDataIndexName(packageName);
+    } else {
+      // FIXME:PT define once we have new index patterns (defend workflows team issue #6553)
+      this.fileMetaIndex = getFileMetadataIndexName(packageName);
+      this.fileDataIndex = getFileDataIndexName(packageName);
+    }
```

Comment:
(please see the last [comment](https://github.com/elastic/kibana/pull/157658#discussion_r1196755758))

Reply:
I don't think the additional `else {}` is needed since typescript already ensure that `type` must be one of the two values (yes, we can bypass that - but then a dev is shooting themselves in the foot.